pkg/start/start.go

```diff
@@ -133,19 +135,28 @@ func (o *Options) Run(ctx context.Context) error {
 	// check to see if techpreview should be on or off.  If we cannot read the featuregate for any reason, it is assumed
 	// to be off.  If this value changes, the binary will shutdown and expect the pod lifecycle to restart it.
 	startingFeatureSet := ""
-	gate, err := cb.ClientOrDie("feature-gate-getter").ConfigV1().FeatureGates().Get(ctx, "cluster", metav1.GetOptions{})
-	switch {
-	case apierrors.IsNotFound(err):
-		// if we have no featuregates, then the cluster is using the default featureset, which is "".
-		// This excludes everything that could possibly depend on a different feature set.
-		startingFeatureSet = ""
-	case err != nil:
-		// client-go automatically retries network blip errors on GETs for 30s by default.  If we fail longer than that
-		// the operator won't be able to do work anyway.  Return the error and crashloop.
-		return err
 
-	default:
-		startingFeatureSet = string(gate.Spec.FeatureSet)
+	if err := wait.PollImmediate(2*time.Second, 30*time.Second, func() (done bool, err error) {
+		gate, err := cb.ClientOrDie("feature-gate-getter").ConfigV1().FeatureGates().Get(ctx, "cluster", metav1.GetOptions{})
+		switch {
+		case apierrors.IsNotFound(err):
+			// if we have no featuregates, then the cluster is using the default featureset, which is "".
+			// This excludes everything that could possibly depend on a different feature set.
+			startingFeatureSet = ""
+			return true, nil
+			// retry some network blips that client-go does not, to survive brief API disruptions during install
+		case net.IsTimeout(err) || net.IsConnectionRefused(err):
+			return false, nil
+		case err != nil:
+			// client-go automatically retries some network blip errors on GETs for 30s by default.  If we fail longer than that
+			// the operator won't be able to do work anyway.  Return the error and crashloop.
```

Comment:
This comment doesn't make as much sense now.  Abu pointed out [the client-go retry logic][1] here, and maybe that is giving us 30s of retries under connection resets or probable-EOFs.  But with the wrapping `PollImmediate`, we no longer care about what client-go does, because our poller will retry if client-go fails.  Can we drop this comment?

[1]: https://github.com/kubernetes/kubernetes/blob/93d1c43b491a39740297cf1e76782b99d8739865/staging/src/k8s.io/client-go/rest/request.go#L940-L952

Reply:
I only made it retry on `net.IsTimeout(err) || net.IsConnectionRefused(err)`, see L148. For any other error we propagate it out which *ends* the polling. Do you thing we should just retry everything?

And if yes, maybe the timeout in `PollImmediate` should be lower, like 25s, so that we either retry short blips ourselves, or we hit the client-go retry cases exactly once.

Reply:
addressed in 296ae5f8a950fbda3e348dd131bde3d3a95f92ce